core/event/event-manager.js

```diff
@@ -1167,12 +1185,27 @@ var EventManager = exports.EventManager = Montage.specialize(/** @lends EventMan
 
                 if (bubbles) {
                     // TODO why on the document and not the window?
-                    return /* isWindow*/target.screen ? target.document : target.ownerDocument;
+                    var shadowRoot;
+                    return /* isWindow*/target.screen ? target.document :
+                        (shadowRoot = this.getShawdowRootFromNode(target)) ?
+                            shadowRoot : target.ownerDocument;
                 } else {
                     return target;
                 }
             }
+        }
+    },
 
+    getShawdowRootFromNode: {
+        value: function isInShadow(node) {
+            if (window.ShadowRoot) {
+                while (node) {
+                    if (node.toString() === "[object ShadowRoot]") {
```

Comment:
No better way to test if a node is a ShadowRoot?

Reply:
I haven't found any other safe way.